### PR TITLE
[gpt_oss_d_p] GPT-OSS prefill: package scaffold + attention (GQA/sinks/sliding)

### PR DESCRIPTION
**Pull request by mdragulaTT**
## Problem
GPT-OSS-120B needs a long-context **prefill** on Blackhole Galaxy (TP=8/SP=4/EP=32). The existing `models/demos/gpt_oss` is Wormhole/decode-first and isn't structured to reuse the DeepSeek/MiniMax-M3 prefill stack (EP-MoE dispatch/combine, chunked block-cyclic KV, the `common/prefill` engine).

## Context
First PR of a **stacked chain** that builds a fresh `gpt_oss_d_p` package "the MiniMax-M3 way." This PR is **package scaffold + attention only** — no MoE / KV cache / runtime yet.

Chain: **#50223 (this)** → #50228 (chunked KV+RoPE) → #50254 (MoE) → #50265 (model+runtime+adapter).

## Solution
| Piece | What |
|---|---|
| Package | new `gpt_oss_d_p/`, mirrors `minimax_m3/` layout |
| Attention | GQA (64 Q / 8 KV heads, `head_dim=64`) + full YaRN RoPE + per-Q-head attention sinks + per-layer sliding(128)/full alternation |
| Sinks | stored pre-divided by the softmax scale so the SDPA kernel's internal scaling reproduces HF |
| Multi-device | `MeshConfig`: TP shards KV-heads (1/col), SP shards the sequence. SP>1 uses a **correctness-first AllGather + single-chip SDPA stand-in** (efficient ring-SDPA swaps in later; the stand-in is documented as SP-placeholder and not yet SP-validated) |

## Tests
`tests/unit/test_attention_vs_ref.py` — GQA + sinks + sliding vs a torch reference on a single Blackhole card, both a sliding and a full layer, PCC ≥ 0.99. **2 passed.**

## Next (later PRs in the chain)
Chunked KV+RoPE (#50228) → MoE wrapper (#50254) → model/runtime/adapter (#50265). The AllGather SP stand-in is replaced by the native ring SDPA (sinks + sliding + halo CCL) for efficient sequence-parallel attention.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/gpt-oss-attention)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/gpt-oss-attention)
### CI Notes
The red **Sanity tests** badge above is from a test unrelated to this PR, which adds only a new model
package (`gpt_oss_d_p`) with no ttnn ops or kernels:

- **Sanity tests**: `test_dram_group_norm_vae_welford_reciprocal_performance`, an SDXL VAE group-norm
  performance test on Wormhole that trips its +/-1.5% perf-tolerance window (it measured *faster* than
  the baseline). Flaky on `main`, independent of this branch.

This PR's own coverage is green: the attention unit test passes via the scoped `all-model-tests` run
(`Model[gpt-oss-120b] SKUs[bh_p150]`); **PR Gate** and **Blackhole sanity** are green.

### CODE REVIEW

File: models/demos/gpt_oss_d_p/tt/attention/prefill.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        if batch_size > 1:
+            for b in range(batch_size):
+                k_b = ttnn.slice(
+                    k_for_cache, (b, 0, 0, 0), (b + 1, k_for_cache.shape[1], k_for_cache.shape[2], k_for_cache.shape[3])
+                )
+                v_b = ttnn.slice(
+                    v_for_cache, (b, 0, 0, 0), (b + 1, v_for_cache.shape[1], v_for_cache.shape[2], v_for_cache.shape[3])
+                )
+                ttnn.fill_cache(k_cache, k_b, batch_idx=b)
+                ttnn.fill_cache(v_cache, v_b, batch_idx=b)
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/prefill.py]: Done. Batched writes now use `batch_idx=user_id + b`. (Note: this whole `fill_cache` block is replaced by `write_kv_chunk` in the next PR (#50228); the fix applies to the single-shot path that lives in this PR.)

File: models/demos/gpt_oss_d_p/tt/config.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        self.tp_axis = tp_axis
+        self.ep_axis = 0 if tp_axis == 1 else 1
+        self.sp_axis = self.ep_axis
+        self.total_devices = self.mesh_shape[0] * self.mesh_shape[1]
+        self._validate()
+
+    def _validate(self):
+        tp_dim_size = self.mesh_shape[self.tp_axis]
+        if self.tp > tp_dim_size:
+            raise ValueError(f"TP({self.tp}) > mesh_{self.tp_axis}_size({tp_dim_size})")
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/config.py]: Done. `_validate` now requires `tp == mesh_axis_size` and rejects sub-axis TP, since `shard_mapper` shards the full mesh axis.

File: models/demos/gpt_oss_d_p/tests/unit/test_attention_vs_ref.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    sinks = w["sinks"].view(1, NQ, 1, 1).expand(B, NQ, S, 1)  # [B, NQ, S, 1]
+    combined = torch.cat([scores, sinks], dim=-1)  # [B, NQ, S, S+1]
+    probs = torch.softmax(combined, dim=-1)[..., :S]  # drop the sink column
+
+    out = probs @ v  # [B, NQ, S, HD]
+    out = out.transpose(1, 2).reshape(B, S, NQ * HEAD_DIM)
+    return out @ w["o"].t() + w["o_bias"]
+
+
+@pytest.mark.parametrize("mesh_device", [(1, 1)], indirect=True)
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tests/unit/test_attention_vs_ref.py:149]: Not in this PR. The `gpt_oss_d_p` unit suite is still landing across the stacked PRs (#50223 through #50265), so wiring a single test into the shared CI registry now would need repeated edits as more tests arrive. We will register the whole `models/demos/gpt_oss_d_p/tests/unit/` directory once the suite is complete (bh single-card tier-2, galaxy tier-1 for the EP=32 model test). Tracked as a merge follow-up.

File: models/demos/gpt_oss_d_p/tt/config.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        if self.total_devices % self.tp != 0:
+            raise ValueError(f"TP({self.tp}) does not divide total_devices({self.total_devices})")
+        if (self.mesh_shape, self.tp) != (_VALIDATED_MESH_SHAPE, _VALIDATED_TP):
+            logger.warning(
+                f"MeshConfig(mesh_shape={self.mesh_shape}, tp={self.tp}) is untested — only "
+                f"mesh_shape={_VALIDATED_MESH_SHAPE}, tp={_VALIDATED_TP} (SP=4, EP=4) is the GPT-OSS target."
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/config.py]: Done. The warning and the target comment now report EP=32, matching the README and PR.

File: models/demos/gpt_oss_d_p/README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+## Attention (net-new): shapes & correctness notes
+
+Per chip, per layer, one prefill chunk (`S_loc = S/SP`):
+
+| tensor | shape | dtype | layout |
+|---|---|---|---|
+| Q | `[1, 8, S_loc, 64]` (8 of 64 Q-heads) | bf16 | TILE |
+| K | `[1, 1, S_loc, 64]` (1 of 8 KV-heads) | bf16 (cache bf8_b) | TILE |
+| V | `[1, 1, S_loc, 64]` | bf16 (cache bf8_b) | TILE |
+| sinks | `[8]` (per local Q-head), **pre-scaled by 1/√64** | bf16 | — |
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/README.md]: Done. The table row and the loader comment now say 'pre-divided by `config.scaling`' (the 1/sqrt(head_dim) softmax scale, i.e. multiply by sqrt(64)).

File: models/demos/gpt_oss_d_p/tt/attention/prefill.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    """Single-chip GQA SDPA with sliding-window + attention-sink (the P1 dense path)."""
+    return ttnn.transformer.scaled_dot_product_attention(
+        tt_q,
+        tt_k,
+        tt_v,
+        is_causal=True,
+        sliding_window_size=config.sliding_window,
+        attention_sink=weights.sinks,
+        program_config=program_config.get_prefill_sdpa_config(mesh_device, seq_len),
+        compute_kernel_config=program_config.get_compute_kernel_config(),
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/attention/prefill.py:43]: Done. `_run_sdpa` now passes `scale=config.scaling`. For gpt-oss this equals the kernel default (1/sqrt(head_dim)) so numerics are unchanged, but it honors an explicitly configured scale and matches the gpt_oss decode path.

File: models/demos/gpt_oss_d_p/tests/unit/test_attention_vs_ref.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tests/unit/test_attention_vs_ref.py:149]: Update: actually done in this PR. Registered `models/demos/gpt_oss_d_p/tests/unit/` in `tests/pipeline_reorg/models_unit_tests.yaml` (entry `gpt-oss-120b-dp`, `bh_p150` tier-2, directory glob so the kv-cache/MoE tests auto-run as they land). Verified via a scoped `all-model-tests` dispatch. Single-card units live here; the multi-device / EP=32 galaxy coverage still follows deepseek_v3_d_p into blackhole_e2e / the galaxy job at P5.

**Comment by mtairum** [comment]
what's `d_p` in this context? Data parallel?
Is this part of blaze?

I'm trying to figure out what's the best naming convention for this on the unit tests yaml.

**Comment by mdragulaTT** [comment]
`d_p` stands for disaggregated prefill. It follows the naming convention from `models/demos/deepseek_v3_d_p` (README: "DeepSeek V3: Disaggregated prefill"). The prefill stage runs disaggregated from decode, driven by the model-agnostic prefill engine in `models/demos/common/prefill`.

So this is deliberately a separate package from the existing models/demos/gpt_oss (the Wormhole, decode-first demo): `gpt_oss_d_p` is the prefill-stage implementation for GPT-OSS-120B. It's built mirroring `models/demos/minimax_m3` (which itself came from the GLM / DeepSeek prefill lineage) and reuses the shared DeepSeek EP-MoE dispatch/combine substrate plus the common/prefill runner. It is not part of blaze.

**Review by mtairum — APPROVED** [review]

**Comment by mtairum** [comment]
Gotcha. 

In that case, let's update `tests/pipeline_reorg/models_unit_tests.yaml` to make it more verbose.

`- name: GPT-OSS-120B (disaggregated prefill) unit tests`
I'll approve to unblock.

**Review by roseli-TT — APPROVED** [review]

**Review by jmalone-tt — APPROVED** [review]
